Allow default arguments to be used for measure() method

chadallee opened the pull request:
I want this to work (default parameter value):

```python
@measurementservice.output("my_output", nims.DataType.Double)
@measurementservice.configuration("my_param_1", nims.DataType.Double)
#@measurementservice.configuration("my_param_2_still_debugging_might_not_use_this", nims.DataType.Double)
def measure(my_param_1,
            my_param_2_still_debugging_might_not_use_this = None):
    return (0.0,)
```

<!-- TODO: Check the below box with an 'x' indicating you've read and followed [CONTRIBUTING.md](https://github.com/ni/measurement-plugin-python/blob/main/CONTRIBUTING.md) -->
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-plugin-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Allows default parameters to be used from the `measure()` method for the measurement plugin instead of throwing an exception.

Currently, the above plugin script will throw an exception (and not a particularly helpful one):
> 2025-02-26 14:35:42,820 ERROR: Exception iterating responses: 2
Traceback (most recent call last):
  File ".venv\Lib\site-packages\grpc\_server.py", line 654, in _take_response_from_response_iterator
    return next(response_iterator), True
           ^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv\Lib\site-packages\ni_measurement_plugin_sdk_service\grpc\loggers.py", line 391, in __next__
    response = next(self._inner_iterator)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv\Lib\site-packages\ni_measurement_plugin_sdk_service\_internal\grpc_servicer.py", line 381, in Measure
    mapping_by_variable_name = _get_mapping_by_parameter_name(
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".venv\Lib\site-packages\ni_measurement_plugin_sdk_service\_internal\grpc_servicer.py", line 139, in _get_mapping_by_parameter_name
    mapping_by_variable_name[parameter.name] = mapping_by_id[i]
                                               ~~~~~~~~~~~~~^^^
KeyError: 2

### Why should this Pull Request be merged?

This problem is admittedly easy to workaround (comment out both the `@measurementservice.configuration()` AND the method parameter), but this fixes a corner case that really doesn't have to error -- and it makes the exception that gets thrown when there is no default much better .

### What testing has been done?

My measurement plugin works and uses default parameters for any arguments that weren't specified with `@measurementservice.configuration()`.

Also tested that any arguments **without** a default parameter that also weren't configured with `@measurementservice.configuration()` does cause an exception to be thrown -- and the exception is more useful:

> 2025-02-26 14:23:23,980 ERROR: Exception iterating responses: measure() missing 1 required positional argument: 'my_param_2_still_debugging_might_not_use_this'

bkeryan commented on packages/service/ni_measurement_plugin_sdk_service/_internal/grpc_servicer.py:
Would this also solve your problem?

https://github.com/ni/measurement-plugin-python/issues/335

chadallee replied:
Yes, absolutely. That would be great.

bkeryan commented on packages/service/ni_measurement_plugin_sdk_service/_internal/grpc_servicer.py:
If we're going to allow this use case, we should update one of the test services to use it.

https://github.com/ni/measurement-plugin-python/tree/main/packages/service/tests/utilities/measurements

bkeryan replied:
For example, add something like `extra_param: bool = False` to https://github.com/ni/measurement-plugin-python/blob/main/packages/service/tests/utilities/measurements/loopback_measurement/__init__.py

bkeryan commented on packages/service/ni_measurement_plugin_sdk_service/_internal/grpc_servicer.py:
What happens if the service defines a keyword-only or positional-only parameter?

Perhaps we should filter parameters based on `default` and `kind`.
https://docs.python.org/3/library/inspect.html#inspect.Parameter

chadallee replied:
Interesting! I was not aware of keyword-only or positional-only parameters.

With this change, the behavior of position-only arguments is that you get the following exception:
> 2025-02-27 12:41:16,764 ERROR: Exception iterating responses: measure() got some positional-only arguments passed as keyword arguments: 'In_Boolean'
Traceback (most recent call last):
  File ".\Lib\site-packages\grpc\_server.py", line 654, in _take_response_from_response_iterator
    return next(response_iterator), True
           ^^^^^^^^^^^^^^^^^^^^^^^
  File ".\.venv\Lib\site-packages\ni_measurement_plugin_sdk_service\grpc\loggers.py", line 391, in __next__
    response = next(self._inner_iterator)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".\.venv\Lib\site-packages\ni_measurement_plugin_sdk_service\_internal\grpc_servicer.py", line 389, in Measure
    return_value = self._measure_function(**mapping_by_variable_name)
    
So I guess the alternative you're suggesting would be to filter out that positional-only argument, which would stop the exception from being thrown, but now we're just silently passing nothing to `measure()` function and allowing it to use its default value instead of the actual input.

I feel like throwing the exception seems better in this case. And as you said, using *args would be even better still.

bkeryan replied:
Yeah, I agree filtering out parameters doesn't make sense.

Switching to `*args` would break compatibility with keyword-only parameters.

I think the most compatible approach would be to use `kind` to decide whether to add the parameter to `args` or `kwargs`. I guess that's orthogonal to ignoring function parameters that don't correspond to "configurations", though.

chadallee commented:
> ```
> @measurementservice.configuration("my_param_1", nims.DataType.Double)
> #@measurementservice.configuration("my_param_2_still_debugging_might_not_use_this", nims.DataType.Double)
> def measure(my_param_1,
>             my_param_2_still_debugging_might_not_use_this : None):
> ```
> 
> @chadallee This is a type hint that says the type of `my_param_2_still_debugging_might_not_use_this ` is `None`. Did you mean `= None`?
> 
> There are two types of default parameters in Python and they are often confused.
> 
> 1. The parameter can have a default value.
> 
> ```
> def measure(my_param: str = "", my_param_2: float = 0.0):
> ```
> 
> 2. The parameter's type can accept None in addition to another type.
> 
> ```
> # Python 3.8
> def measure(my_param: typing.Optional[str] = None, my_param_2: typing.Optional[float] = 0.0):
> # Python 3.9+
> def measure(my_param: str | None = None, my_param_2: float | None = None):
> ```
> 
> BTW, this package has its own way of specifying default values, which allows the service to return the default values over gRPC so that they can be displayed in InstrumentStudio or TestStand:
> 
> ```
> @measurement_service.configuration("Float In", nims.DataType.Float, 0.06)
> @measurement_service.configuration("String In", nims.DataType.String, "sample string")
> def measure(
>     float_input: float,
>     string_input: str,
> )
> ```

Oops, good catch. Sorry, I hastily trimmed down the script I was actually testing with and changed the `=` to `:`.
I've fixed my initial PR description

bkeryan commented on packages/service/ni_measurement_plugin_sdk_service/_internal/grpc_servicer.py:
```diff
@@ -134,7 +134,8 @@ def _get_mapping_by_parameter_name(
     mapping_by_variable_name = {}
     for i, parameter in enumerate(signature.parameters.values(), start=1):
-        mapping_by_variable_name[parameter.name] = mapping_by_id[i]
+        if i in mapping_by_id:
+            mapping_by_variable_name[parameter.name] = mapping_by_id[i]
```
You can avoid an extra map lookup here:
```
_UNSET_MAPPING = object()
...
def _get_mapping_by_parameter_name(...):
    ....
    for i, parameter in enumerate(signature.parameters.values(), start=1):
        mapping = mapping_by_id.get(i, _UNSET_MAPPING)
        if mapping is not _UNSET_MAPPING:
            mapping_by_variable_name[parameter.name] = mapping
```
